modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java

```diff
@@ -788,4 +836,269 @@ else if (e.getCause() instanceof IgniteException)
+     * @return Future with cache sizes.
+     */
+    private Future<CacheSize> calcCacheSizeAsync(
+        CacheGroupContext grpCtx,
+        GridDhtLocalPartition locPart
+    ) {
+        return calcExecutor.submit(() -> {
+            try {
+                int grpId = grpCtx.groupId();
+
+                if (failCalcCacheSizeGrpIds.contains(grpId))
+                    return new CacheSize(null, null);
+
+                boolean reserve = false;
+
+                int partId = locPart.id();
+
+                try {
+                    if (!(reserve = locPart.reserve()))
+                        throw new IgniteException("Can't reserve partition");
+
+                    if (locPart.state() != OWNING)
+                        throw new IgniteException("Partition not in state " + OWNING);
+
+                    Map<Integer, Map<String, AtomicLong>> cacheSizeByTbl = new HashMap<>();
+
+                    GridIterator<CacheDataRow> partIter = grpCtx.offheap().partitionIterator(partId);
+
+                    GridQueryProcessor qryProcessor = ignite.context().query();
+                    IgniteH2Indexing h2Indexing = (IgniteH2Indexing)qryProcessor.getIndexing();
+
+                    while (partIter.hasNextX() && !failCalcCacheSizeGrpIds.contains(grpId)) {
+                        CacheDataRow cacheDataRow = partIter.nextX();
+
+                        int cacheId = cacheDataRow.cacheId();
+
+                        GridCacheContext cacheCtx = cacheId == 0 ?
+                            grpCtx.singleCacheContext() : grpCtx.shared().cacheContext(cacheId);
+
+                        if (cacheCtx == null)
+                            throw new IgniteException("Unknown cacheId of CacheDataRow: " + cacheId);
+
+                        if (cacheDataRow.link() == 0L)
+                            throw new IgniteException("Contains invalid partition row, possibly deleted");
+
+                        String cacheName = cacheCtx.name();
+
+                        QueryTypeDescriptorImpl qryTypeDesc = qryProcessor.typeByValue(
+                            cacheName,
+                            cacheCtx.cacheObjectContext(),
+                            cacheDataRow.key(),
+                            cacheDataRow.value(),
+                            true
+                        );
+
+                        if (isNull(qryTypeDesc))
+                            continue; // Tolerate - (k, v) is just not indexed.
+
+                        String tableName = qryTypeDesc.tableName();
+
+                        GridH2Table gridH2Tbl = h2Indexing.schemaManager().dataTable(cacheName, tableName);
+
+                        if (isNull(gridH2Tbl))
+                            continue; // Tolerate - (k, v) is just not indexed.
+
+                        cacheSizeByTbl.computeIfAbsent(cacheCtx.cacheId(), i -> new HashMap<>())
+                            .computeIfAbsent(tableName, s -> new AtomicLong()).incrementAndGet();
```

Comment:
Is it enough this checking that say: "This entry presented in index."?
What did happen if indexed column added dynamical?

Reply:
The check will pass correctly, here are tests that will confirm this.
org.apache.ignite.util.GridCommandHandlerIndexingCheckSizeTest#testCheckCacheSizeWhenBrokenCacheWithDynamicAddColumnAndIndex
org.apache.ignite.util.GridCommandHandlerIndexingCheckSizeTest#testCheckCacheSizeWhenBrokenIdxWithDynamicAddColumnAndIndex